### ISSUE TITLE
Feature : (Diagnostic) Automated FSDP Compute–Communication Overlap Diagnostic in Profiler

### ISSUE THREAD

**Issue by pbpatre** [pytorch/torchtune#2954]
[RFC] Feature: Automated FSDP Compute--Communication Overlap Diagnostic in Profiler
==================================================================================

Problem
-------

When training with FSDP on multi‑GPU, users often see low MFU without realizing they are communication‑bound. Diagnosing this today requires downloading large traces and manually inspecting CUDA streams for NCCL bottlenecks.

Proposed Solution
-----------------

Add a lightweight `FSDPOverlapAnalyzer` hooked into `setup_torch_profiler`'s `on_trace_ready`. After each profiling cycle, it parses `prof.key_averages()` and computes a heuristic overlap efficiency based on aggregate NCCL vs compute time. It logs a compact summary to the logger (rank‑0 only). Example:

```
--- FSDP Overlap Efficiency Report ---
Total Compute Time: 28.86ms
Total NCCL Time: 192.61ms
Network Tax (Bubble): 163.75ms
Overlap Efficiency (heuristic): 13.03%
Status: WARNING (Communication Bound)

```

Configuration
-------------

Add an opt‑in flag:

yaml

```
profiler:
  analyze_fsdp_overlap: true
```

Alternatives Considered
-----------------------

External tooling (Holistic Trace Analysis), but that introduces heavy dependencies. This can be done using existing `key_averages()` output.

Limitations
-----------

This is a heuristic derived from aggregated event totals and does not compute true overlap windows. It's intended as a fast, lightweight diagnostic, not a full trace replacement.

Context
-------

I hit this while scaling a 3B model on 2x H200s. Disabling NVLink created clear NCCL stalls that were only obvious in the trace. Having this metric printed automatically would save hours of debugging. I have a working prototype and can submit a PR.

**Comment by pbpatre** [comment]
Following up on this—I've noticed the project’s recent shift in focus toward the torchtitan ecosystem for high-scale, production-ready training.

Given that this overlap diagnostic is most critical for the massive-scale workloads (70B+ parameters) that torchtitan targets, would the maintainers prefer I open an RFC and PR there instead? My goal is to land this where it provides the highest architectural signal for the community.
